HPCC-17037 Add polling option to ecl command line "run"

afishbeck opened the pull request:
ecl run <target> <ecl> --poll

Rather than connecting to the server and blocking indefinately will now
poll for changes to the workunit state.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 342073c3f7b93cc627c35e77df000cac6acc5374
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.532 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@afishbeck Who can review this?

afishbeck commented:
@wangkx can you review?

wangkx reviewed (commented):
@afishbeck only 2 questions.

afishbeck commented on esp/services/ws_workunits/ws_workunitsService.cpp:
```diff
@@ -1052,6 +1052,58 @@ bool CWsWorkunitsEx::onWURun(IEspContext &context, IEspWURunRequest &req, IEspWU
+    {
+        StringBuffer wuid = req.getWuid();
+        WsWuHelpers::checkAndTrimWorkunit("WUFullResult", wuid);
+
```
Good catch.

afishbeck commented on esp/services/ws_workunits/ws_workunitsService.cpp:
```diff
@@ -1052,6 +1052,58 @@ bool CWsWorkunitsEx::onWURun(IEspContext &context, IEspWURunRequest &req, IEspWU
+                break;
+            }
+            default:
+                if (!cw)
```
For fun?  I'll remove it.

afishbeck committed:
HPCC-17037 Add polling option to ecl command line "run"

ecl run <target> <ecl> --poll

Rather than connecting to the server and blocking indefinately will now
poll for changes to the workunit state.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@wangkx updated with both of your suggestions.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 4eb6559c91179eabba0ba259763c89393a9eca73
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 95.201 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented:
@afishbeck looks fine now.

richardkchapman commented on ecl/eclcmd/eclcmd_core.cpp:
```diff
@@ -604,7 +823,7 @@ class EclCmdRun : public EclCmdWithEclTarget
         else if (optTargetCluster.length())
             req->setCluster(optTargetCluster.get());
-        req->setWait((int)optWaitTime);
+        req->setWait((int)(optPoll ? 0 : optWaitTime));
```
I was thinking that even in poll mode, you would wait for (say) 1 second the first time you issue the call, so that subsecond executions did not incur any additional delay. Not a massive deal if that's too complex, but...

afishbeck replied:
Yes before I made the polling an explicit user option that was my intent.  No reason not do do it even now.  I'll make the change.  I may make the initial wait even longer, say 5 or 10 secs?

afishbeck committed:
HPCC-17037 ecl-run wait connected for first 10 seconds even when polling

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@richardkchapman updated with initial connected wait, even when polling.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 7a29d52b9fd07316c12ecb509e95059584078523
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.664 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



HPCC Stop: OK
HPCC Uninstall: OK
